src/goose/cli/session.py

```diff
@@ -150,23 +157,41 @@ def run(self) -> None:
         """
         message = self.process_first_message()
         while message:  # Loop until no input (empty string).
-            self.notifier.start()
-            try:
-                self.exchange.add(message)
-                self.reply()  # Process the user message.
-            except KeyboardInterrupt:
-                self.interrupt_reply()
-            except Exception:
-                # rewind to right before the last user message
-                self.exchange.rewind()
-                print(traceback.format_exc())
-                print(
-                    "\n[red]The error above was an exception we were not able to handle.\n\n[/]"
-                    + "These errors are often related to connection or authentication\n"
-                    + "We've removed the conversation up to the most recent user message"
-                    + " - [yellow]depending on the error you may be able to continue[/]"
-                )
-            self.notifier.stop()
+            span_attributes = {
+                "goose.role": message.role,
+                "goose.id": message.id,
+            }
+
+            # For starters, only add to the trace the first text content
+            first_content = message.content[0] if message.content else None
+            if isinstance(first_content, Text):
+                span_attributes["goose.text"] = first_content.text
+
+            with self.tracer.start_as_current_span(message.role, attributes=span_attributes) as span:
+                self.notifier.start()
+                try:
+                    self.exchange.add(message)
+                    self.reply()  # Process the user message.
+                except KeyboardInterrupt:
+                    # TODO: should we make interrupting an error? If not, how should we mark this?
+                    # span as it was interrupted?
+                    span.set_status(OtelStatus(OtelStatusCode.ERROR, "KeyboardInterrupt"))
+                    self.interrupt_reply()
+                except Exception as e:
+                    # rewind to right before the last user message
+                    self.exchange.rewind()
+                    span.set_status(OtelStatus(OtelStatusCode.ERROR, str(e)))
+                    span.record_exception(e)
+                    print(traceback.format_exc())
+                    print(
+                        "\n[red]The error above was an exception we were not able to handle.\n\n[/]"
+                        + "These errors are often related to connection or authentication\n"
+                        + "We've removed the conversation up to the most recent user message"
+                        + " - [yellow]depending on the error you may be able to continue[/]"
+                    )
+                finally:
+                    self.notifier.stop()
+                    span.end()
```

Comment:
Oh just noticed, we don't need to explicitly end since we don't set `end_on_exit = False` when `start_as_current_span`. I think in practice double-end is a no-op but some strict leak detector type of validators might not like it if ever implemented

Reply:
another good catch, actually I saw the leak just now, so you were right.